### ISSUE TITLE
2.2.0 java.lang.ClassNotFoundException: com.stripe.cots.CotsAdapterProvider

### ISSUE THREAD

**Issue by matejkramny** [stripe/stripe-terminal-android#155]
## Summary
This error (title) appears when initializing the Terminal SDK.

We are also having trouble finding any readers (in debug mode), but this seems to work on the Example app, so that will be something with our implementation. I'm mentioning this because it sounds related, if something is not being included in the build or this exception stops the SDK from working.

Might be related to #142

## Code to reproduce
Run the example app (Kotlin)

You may have to comment out the VmPolicy as it detects a leak (due to stack trace below) and kills the app.

## Android version
11

## Impacted devices (Android devices or readers)
all

## SDK version
2.2.0

## Other information
```
W/System.err: java.lang.ClassNotFoundException: com.stripe.cots.CotsAdapterProvider
        at java.lang.Class.classForName(Native Method)
        at java.lang.Class.forName(Class.java:454)
        at java.lang.Class.forName(Class.java:379)
        at com.stripe.stripeterminal.dagger.CotsModule.provideCotsAdapter(CotsModule.kt:46)
        at com.stripe.stripeterminal.dagger.CotsModule_ProvideCotsAdapterFactory.provideCotsAdapter(CotsModule_ProvideCotsAdapterFactory.java:39)
        at com.stripe.stripeterminal.dagger.CotsModule_ProvideCotsAdapterFactory.get(CotsModule_ProvideCotsAdapterFactory.java:29)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.stripe.stripeterminal.dagger.CotsModule_ProvideCotsProxyAdapterFactory.get(CotsModule_ProvideCotsProxyAdapterFactory.java:49)
        at com.stripe.stripeterminal.dagger.CotsModule_ProvideCotsProxyAdapterFactory.get(CotsModule_ProvideCotsProxyAdapterFactory.java:14)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.stripe.stripeterminal.dagger.TerminalModule_ProvideProxyAdapterFactory.get(TerminalModule_ProvideProxyAdapterFactory.java:55)
        at com.stripe.stripeterminal.dagger.TerminalModule_ProvideProxyAdapterFactory.get(TerminalModule_ProvideProxyAdapterFactory.java:17)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.stripe.stripeterminal.TerminalSession_Factory.get(TerminalSession_Factory.java:65)
        at com.stripe.stripeterminal.TerminalSession_Factory.get(TerminalSession_Factory.java:15)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.stripe.stripeterminal.Terminal_Factory.get(Terminal_Factory.java:31)
        at com.stripe.stripeterminal.Terminal_Factory.get(Terminal_Factory.java:9)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.stripe.stripeterminal.dagger.DaggerTerminalComponent.getTerminal(DaggerTerminalComponent.java:741)
        at com.stripe.stripeterminal.Terminal$Companion.initTerminal(Terminal.kt:811)
        at com.stripe.example.MainActivity.initialize(MainActivity.kt:273)
        at com.stripe.example.MainActivity.onResume(MainActivity.kt:77)
        at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1456)
        at android.app.Activity.performResume(Activity.java:8129)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4428)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4470)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7664)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
    Caused by: java.lang.ClassNotFoundException: Didn't find class "com.stripe.cots.CotsAdapterProvider" on path: DexPathList[[zip file "/data/app/~~CxodMzqbtLwWaW-a2QKqWQ==/com.stripe.example-3UrakJSMmBdLOsk_pv_cHg==/base.apk"],nativeLibraryDirectories=[/data/app/~~CxodMzqbtLwWaW-a2QKqWQ==/com.stripe.example-3UrakJSMmBdLOsk_pv_cHg==/lib/arm64, /data/app/~~CxodMzqbtLwWaW-a2QKqWQ==/com.stripe.example-3UrakJSMmBdLOsk_pv_cHg==/base.apk!/lib/arm64-v8a, /system/lib64, /system_ext/lib64, /product/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:207)
W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
    	... 37 more
```

**Comment by dhenry-stripe** [comment]
Thanks for reporting @matejkramny.

> You may have to comment out the VmPolicy as it detects a leak (due to stack trace below) and kills the app.

The error you posted is misleading. It's an expected error and isn't responsible for crashing the app. That's more likely caused by not setting `ApiClient.BACKEND_URL` or disabling location permissions.

We shouldn't be leaking the stacktrace though. I'll leave this issue open to address that.

**Comment by matejkramny** [comment]
Got it.

I set the backend API though and it kept mentioning something about a leak.
Might be another red herring

**Comment by rscherf** [comment]
I'm also receiving this after upgrading to `2.2.0` from `b3` in my Java app. I'm not using an implementation like the example app with `ApiClient`, however. Not sure if this is helpful information, but here is the stack trace when in DEBUG attempting to find simulators.

```
D/StripeTerminal: class=Terminal app_language=KOTLIN
D/StripeTerminal: class=LocationHandler message=startListening
W/System.err: java.lang.ClassNotFoundException: com.stripe.cots.CotsAdapterProvider
        at java.lang.Class.classForName(Native Method)
        at java.lang.Class.forName(Class.java:454)
        at java.lang.Class.forName(Class.java:379)
        at com.stripe.stripeterminal.dagger.CotsModule.provideCotsAdapter(CotsModule.kt:46)
W/System.err:     at com.stripe.stripeterminal.dagger.CotsModule_ProvideCotsAdapterFactory.provideCotsAdapter(CotsModule_ProvideCotsAdapterFactory.java:39)
W/System.err:     at com.stripe.stripeterminal.dagger.CotsModule_ProvideCotsAdapterFactory.get(CotsModule_ProvideCotsAdapterFactory.java:29)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.stripe.stripeterminal.dagger.CotsModule_ProvideCotsProxyAdapterFactory.get(CotsModule_ProvideCotsProxyAdapterFactory.java:49)
W/System.err:     at com.stripe.stripeterminal.dagger.CotsModule_ProvideCotsProxyAdapterFactory.get(CotsModule_ProvideCotsProxyAdapterFactory.java:14)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.stripe.stripeterminal.dagger.TerminalModule_ProvideProxyAdapterFactory.get(TerminalModule_ProvideProxyAdapterFactory.java:55)
W/System.err:     at com.stripe.stripeterminal.dagger.TerminalModule_ProvideProxyAdapterFactory.get(TerminalModule_ProvideProxyAdapterFactory.java:17)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
W/System.err:     at com.stripe.stripeterminal.TerminalSession_Factory.get(TerminalSession_Factory.java:65)
        at com.stripe.stripeterminal.TerminalSession_Factory.get(TerminalSession_Factory.java:15)
W/System.err:     at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at com.stripe.stripeterminal.Terminal_Factory.get(Terminal_Factory.java:31)
W/System.err:     at com.stripe.stripeterminal.Terminal_Factory.get(Terminal_Factory.java:9)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
W/System.err:     at com.stripe.stripeterminal.dagger.DaggerTerminalComponent.getTerminal(DaggerTerminalComponent.java:741)
        at com.stripe.stripeterminal.Terminal$Companion.initTerminal(Terminal.kt:811)
        at com.stripe.stripeterminal.Terminal.initTerminal(Unknown Source:2)
W/System.err:     at nudge.payment.activities.BaseAuthorizedActivity.initializeTerminal(BaseAuthorizedActivity.java:244)
W/System.err:     at nudge.payment.activities.BaseAuthorizedActivity.doConnectReader(BaseAuthorizedActivity.java:234)
W/System.err:     at nudge.payment.activities.AmountActivity.selectDrawerItem(AmountActivity.java:211)
        at nudge.payment.activities.BaseAuthorizedActivity$16.onNavigationItemSelected(BaseAuthorizedActivity.java:555)
        at com.google.android.material.navigation.NavigationView$1.onMenuItemSelected(NavigationView.java:217)
W/System.err:     at androidx.appcompat.view.menu.MenuBuilder.dispatchMenuItemSelected(MenuBuilder.java:834)
W/System.err:     at androidx.appcompat.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:158)
        at androidx.appcompat.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:985)
W/DefaultDispatch: type=1400 audit(0.0:134): avc: denied { read } for name="cache" dev="dm-4" ino=17 scontext=u:r:untrusted_app_29:s0:c148,c256,c512,c768 tcontext=u:object_r:cache_file:s0 tclass=lnk_file permissive=0 app=com.nudge.payment
W/System.err:     at com.google.android.material.internal.NavigationMenuPresenter$1.onClick(NavigationMenuPresenter.java:416)
        at android.view.View.performClick(View.java:7441)
W/System.err:     at android.view.View.performClickInternal(View.java:7418)
        at android.view.View.access$3700(View.java:835)
W/System.err:     at android.view.View$PerformClick.run(View.java:28674)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loopOnce(Looper.java:201)
W/System.err:     at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7822)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
    Caused by: java.lang.ClassNotFoundException: Didn't find class "com.stripe.cots.CotsAdapterProvider" on path: DexPathList[[dex file "/data/data/com.nudge.payment/code_cache/.overlay/base.apk/classes7.dex", zip file "/data/app/~~hGPO7f4obtcXdfid4AWRyw==/com.nudge.payment--YWlyfMDbXTPl3fCp-w7xg==/base.apk"],nativeLibraryDirectories=[/data/app/~~hGPO7f4obtcXdfid4AWRyw==/com.nudge.payment--YWlyfMDbXTPl3fCp-w7xg==/lib/x86_64, /data/app/~~hGPO7f4obtcXdfid4AWRyw==/com.nudge.payment--YWlyfMDbXTPl3fCp-w7xg==/base.apk!/lib/x86_64, /system/lib64, /system_ext/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:218)
W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
W/System.err:   ... 43 more
D/StripeTerminal: class=Terminal device_rooted=TRUE
D/StripeTerminal: class=ConnectionTokenManager message=refreshToken
I/StripeTerminal: class=TransactionStateMachine message="onStateChanged: null -> EMPTY: null"
I/StripeTerminal: class=TransactionManager message="handleCardStatus NO_CARD"
D/StripeTerminal: class=Terminal message=discoverReaders
```

**Comment by dhenry-stripe** [comment]
Thanks @rscherf, you can ignore this output for now. We'll remove it in the next release.

**Comment by dhenry-stripe** [comment]
Fixed in version [2.3.0](https://github.com/stripe/stripe-terminal-android/blob/master/CHANGELOG.md#230---2021-09-21)
